File: code/ch12/ruby/characterization_tests.rb
# §12.6.2 Characterization Tests — probe, promote the observed value, probe an edge.
# The book's step-1 probe asserts the deliberately wrong "XXX" and fails with:
#   Expected: "XXX"
#     Actual: "E10"
# Here the probe is wrapped in assert_raises so the suite stays green.
# Runs standalone: ruby characterization_tests.rb
require "minitest/autorun"

def legacy_fee_code(visit_type)                     # inherited: no docs, no tests
  { "exam" => "E10", "lab" => "L20", "vaccine" => "V30" }.fetch(visit_type, "E10")
end

class TestFeeCode < Minitest::Test
  def test_probe_fails_as_intended                  # the book shows the raw probe failing
    failure = assert_raises(Minitest::Assertion) do
      assert_equal "XXX", legacy_fee_code("phone")  # deliberately wrong
    end
    assert_includes failure.message, 'Expected: "XXX"'
    assert_includes failure.message, 'Actual: "E10"'
  end

  def test_unknown_type_bills_as_exam               # observed value, promoted
    assert_equal "E10", legacy_fee_code("phone")
  end

  def test_empty_type_bills_as_exam                 # edge probe: pinned, bug or not
    assert_equal "E10", legacy_fee_code("")
  end
end

File: code/ch12/python/characterization_tests.py
# §12.6.2 Characterization Tests — probe, promote the observed value, probe an edge.
# Runs standalone: python3 characterization_tests.py
# Under pytest, the step-1 probe fails by design with the line the book shows:
#   AssertionError: assert 'E10' == 'XXX'

def legacy_fee_code(visit_type):                    # inherited: no docs, no tests
  return {"exam": "E10", "lab": "L20", "vaccine": "V30"}.get(visit_type, "E10")

def test_probe_unknown_type():
  assert legacy_fee_code("phone") == "XXX"        # deliberately wrong

def test_unknown_type_bills_as_exam():              # observed value, promoted
  assert legacy_fee_code("phone") == "E10"

def test_empty_type_bills_as_exam():                # edge probe: pinned, bug or not
  assert legacy_fee_code("") == "E10"

if __name__ == "__main__":
  try:
    test_probe_unknown_type()
  except AssertionError:
    print("probe failed as intended; observed value:", legacy_fee_code("phone"))
  test_unknown_type_bills_as_exam()
  test_empty_type_bills_as_exam()
  print("2 promoted characterization tests pass")
